### ISSUE TITLE
bug: no-event-handler (0.10.1): false positives on props passed to React Query options and on hook args read inside synchronization effects

### ISSUE THREAD

**Issue by nachogarcia** [nickjvandyke/eslint-plugin-react-you-might-not-need-an-effect#70]
### Did you check docs and existing issues?

- [x] I have read the plugin docs
- [x] I have searched the existing issues
- [x] I have read [You Might Not Need An Effect](https://react.dev/learn/you-might-not-need-an-effect) and am reasonably confident that I *do* need an effect

### Plugin version

0.10.1

### ESLint version

N/A — reproduced via oxlint 1.63.0's ESLint-plugin bridge

### Describe the bug

After bumping from `0.10.0` → `0.10.1` (which extended `no-event-handler` to flag props — commit [`d246668`](https://github.com/nickjvandyke/eslint-plugin-react-you-might-not-need-an-effect/commit/d246668)), the rule started firing on two patterns that don't look like the documented anti-pattern. The docs describe the anti-pattern as *"using state and an effect as an event handler"* (e.g. running `showNotification` inside an effect that reacts to `product.isInCart`), but the new behavior fires in cases that aren't event-handler-shaped at all.

### False positive A — props read inside React Query options object literals

The rule flags a prop/argument when it's read as a property of an options object passed to `useQuery` / `useSuspenseQuery` / a query-options factory.

```tsx
import { useQuery } from '@tanstack/react-query'

function useThing({ id, enabled }: { id: string; enabled: boolean }) {
  return useQuery({
    queryKey: ['thing', id],
    queryFn: () => fetch(`/api/thing/${id}`).then(r => r.json()),
    enabled, // ← flagged: "Avoid using props and effects as an event handler"
  })
}
```

`useQuery` is not a `useEffect` — passing `enabled` to it is just configuration. The rule appears to treat any object-literal property read as if it were inside an effect callback.

Same shape with nested options:

```tsx
function ChatPanel({ sessionId }: { sessionId: string }) {
  return useChatbotGetConversationHistoryQuery({
    path: { history_id: sessionId }, // ← `sessionId` flagged
  })
}
```

### False positive B — hook argument read inside an effect that synchronizes with an external store

When a `useEffect` reads a hook argument purely to push it into an external system (localStorage, cookie, URL, etc.), the rule flags the read even though there's no parent callback being invoked and no derived state.

```tsx
import { useEffect } from 'react'

// Captures an optional URL search param and persists it to localStorage + cookie.
// First-touch attribution: never overwrites an existing code.
export function useSaveReferralCode(refCode: string | undefined) {
  useEffect(() => {
    const valid = validateReferralCode(refCode) // ← `refCode` flagged
    if (valid) saveReferredByCode(valid)        // → localStorage + cookie
  }, [refCode])
}
```

Per the React docs section the rule links to ([Sharing logic between event handlers](https://react.dev/learn/you-might-not-need-an-effect#sharing-logic-between-event-handlers)), the anti-pattern is using an effect to do something that should run on a *user event*. Here the trigger is a URL search param showing up after a redirect — there's no user event to hook into in the consuming component, so an effect *is* the right primitive.

Same shape, different external system (Stripe Checkout triggered from a URL param after sign-up redirect):

```tsx
useEffect(() => {
  if (mode !== 'app' || !priceId || hasTriggeredCheckout.current) return // ← `mode`, `priceId` flagged
  hasTriggeredCheckout.current = true
  startCheckout(priceId)
}, [mode, priceId, /* … */])
```

### Expected Behavior

`no-event-handler` should not fire when:

1. The prop/argument is being read as a property of an options object passed to a hook other than `useEffect` / `useLayoutEffect` / `useInsertionEffect` (in particular, it should not treat `useQuery`, `useSuspenseQuery`, `useMutation`, etc. as effect callbacks).
2. The prop/argument is read inside an effect whose body's only side-effect is writing to an external store (localStorage, sessionStorage, cookies, history API, third-party SDKs) and which does *not* call a prop callback or set state derived from the prop.

**Comment by nickjvandyke** [comment]
Thanks for the detailed report!

I think case A is an Oxlint compatibility bug? I added your example as an ESLint test case and get zero warnings. The rule explicitly iterates over `IfStatement` nodes - which your example clearly has none of - so I wonder if Oxlint falsely reports some nodes there?

Re B: While the docs only mention user events, I'd like to catch _all_ event-handling if possible. Imo the same concept applies - code is easier to trace when the event-handling code occurs next to the event, without a state change intermediary. However you make a good point - when its state, we _know_ it's set internally. When it's a prop, we can't be sure. While the URL is external state, it's often set by app code, and this rule applies. However if say the server sets the URL in a redirect (which sounds like your case), then yeah this is incorrect. I'll have to think about which tradeoff is preferrable.

**Comment by ibrahimn-flwls** [comment]
Hitting a closely related shape — same family as your case B, but the rule reaches out of the effect through a custom data-fetching hook and reports on the hook call's arguments, far from the effect itself.

```tsx
function useGetShot(projectId?: string, shotId?: string) {
  const [{ error }] = useQuery({ query: GET_SHOT, variables: { projectId, shotId } });
  return { error };
}

function ShotDetails({ projectId, shotId }: { projectId: string; shotId: string }) {
  const { error } = useGetShot(projectId, shotId);
  //                            ^^^^^^^^^  ^^^^^^  ← both flagged here

  useEffect(() => {
    if (error) showErrorModal();
  }, [error]);

  return null;
}
```

`projectId`/`shotId` are never referenced inside the effect — only `error` is. The rule traces:

`if (error)` → upstream from `error` → out through the `useGetShot(...)` return value → into the call's arguments → lands on the props.

This is structurally different enough from B that the fix-direction (1) you proposed — "stop following return values through custom-hook calls" — covers it cleanly, whereas heuristics that look at "what the effect body does" wouldn't catch it (the effect body doesn't even mention the flagged identifiers).

Same false-positive logic applies to any `useAppSelector(s => selectX(s, propId))` chain, RTK Query / SWR / React Query custom wrappers, etc. — anywhere a custom hook's return value is later read in an effect's `if` test.

**Comment by nickjvandyke** [comment]
> Same false-positive logic applies to any useAppSelector(s => selectX(s, propId)) chain, RTK Query / SWR / React Query custom wrappers, etc. — anywhere a custom hook's return value is later read in an effect's if test.

This is intended for state, but I recently removed it for props in https://github.com/nickjvandyke/eslint-plugin-react-you-might-not-need-an-effect/issues/62. State is a little more reliable because it's never a callback. However it may still be more noise than signal. And easier to think about if it were consistent with props behavior.

**Comment by nickjvandyke** [comment]
Should be fixed in 10.0.3! Lmk if not.
